### PR TITLE
warn when lhost set to 127.0.0.1

### PR DESCRIPTION
**Pull request by bermannoah**
See issue #7188. This pull request adds an alert if you set LHOST to a loopback address (e.g. 127.0.01). Please let me know if there is any additional behavior desired or if I ought to change the warning message. I set the alert to a `print_warning` as requested.

## Verification

The steps needed to make sure this thing works:

- [ ] `./msfconsole`
- [ ] `use exploit/multi/handler`
- [ ] `set payload android/meterpreter/reverse_tcp`
- [ ] `set lhost 127.0.0.1`
- [ ] `set lport 4444`
- [ ] `exploit` 
- [ ] Verify: It will still set `lhost` to `127.0.0.1` but it will print a warning: "You are attempting to listen on a loopback address by setting LHOST to 127.0.0.1, did you mean to set ReverseListenerBindAddress instead?" 
- [ ] `set lhost 0.0.0.0`
- [ ] `export`
- [ ] Verify: It will set `lhost` to that address with no warnings. (Unless you use a different address that produces a warning.)
- [ ] `rake spec` to test that all tests still pass.
### Console Output

```
msf > exploit 
[ ! ] You are attempting to listen on a loopback address by setting LHOST to 127.0.0.1, did you mean to set ReverseListenerBindAddress instead?

[ * ] Started reverse TCP handler on 127.0.0.1:4444
[ * ] Starting the payload handler...
```

### CODE REVIEW

**Comment by busterb** [comment]
thanks @bermannoah . I think we may want to move this validation somewhere else. It seems odd to have this in the middle of the command dispatcher, since you can set LHOST from various places.

File: lib/msf/ui/console/command_dispatcher/core.rb [hunk 1]
```diff
@@ -1541,6 +1541,14 @@ def cmd_set(*args)
       @cache_payloads = nil
     end
 
+    # Raise an alert if LHOST is set to 127.0.0.1
+    name = args[0]
+    ip = args[1]
+
+    if (name.upcase == "LHOST" && ip == "127.0.0.1")
+      print_error("You may experience errors with this choice of address for LHOST.")
```
Comment (wvu-r7) [inline lib/msf/ui/console/command_dispatcher/core.rb]: Wherever we decide to put this check, if `print_warning` is available, please use that instead. Thanks.

Comment (bermannoah) [inline lib/msf/ui/console/command_dispatcher/core.rb]: Will do! Thanks for letting me know.

**Comment by bermannoah** [comment]
@busterb: is there any place in particular you'd recommend moving it? I tried to find a more widely available place to put it and this seemed the most... succinct?

File: lib/msf/ui/console/command_dispatcher/core.rb (hunk 1 above)
Comment (bermannoah) [inline lib/msf/ui/console/command_dispatcher/core.rb]: @wvu-r7: made the change -- thanks!

**Comment by rwhitcroft** [comment]
Might fit better here, right before the bind: https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/handler/reverse.rb#L72

**Comment by rwhitcroft** [comment]
Maybe somewhere in here? https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/handler/reverse.rb#L90

**Comment by busterb** [comment]
hi @bermannoah , I would suggest instead putting the warning inside of lib/msf/core/handler/reverse.rb inside of setup_handler right after it successfully sets up a listening socket, so it warns when it actually starts listening, rather than assignment from the UI. That will also allow it to warn from other contexts as well.

That also puts the warning where the actual networking code lives, so you can do something smart around ReverseListenerBindAddress. You should also maybe check for 127.0.0.0/8 and ::1 instead of just the single IPv4 loopback address.

**Comment by rwhitcroft** [comment]
Might need to go in `reverse_http.rb` as well to handle http(s) stuff. Is there a higher base class I'm not seeing?

**Comment by busterb** [comment]
I think you're right @rwhitcroft.

**Comment by busterb** [comment]
I think adding the test here is appropriate so you catch all of the loopback IPs: https://github.com/rapid7/metasploit-framework/pull/8063/files#r104604972

I also think I see the issue here this is actually warning about. If you're port forwarding over a tunnel, you might want to set ReverseListenerBindAddress to 127.0.0.1, but leave LHOST (which gets embedded into your payload) as an externally routable IP.

So, the message might say 'You are attempting to listen on a loopback address by setting LHOST to #{address}, did you mean to set ReverseListenerBindAddress?' .

**Comment by itsmeroy2012** [comment]
I have a solution to this but I don't know how to pull @bermannoah 's fork as a remote, and send him a pull request on his topic branch.

**Comment by bermannoah** [comment]
@itsmeroy2012 you can pull down my [repo](https://github.com/bermannoah/metasploit-framework) and push to the `lhost-setting-warning` branch or just do the work on your own repo and make a separate PR. I had a couple IRL things come up this weekend and haven't finished fixing my own patch. :)

**Comment by itsmeroy2012** [comment]
I have done the work on my repo. @busterb  told me to make suggestions on here rather than hijacking with an alternate PR and I'm new to github so I'm unable to pull down your repo and push to the 'lhost-setting-warning' branch. Could you guide me through it?

**Comment by bermannoah** [comment]
@itsmeroy2012 yup! Here's a high level overview, let me know if it doesn't match your system setup. Basically:

1 - `git clone git@github.com:bermannoah/metasploit-framework.git` if you're using SSH with github (here's a [guide](https://help.github.com/articles/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent/)) otherwise, `git clone https://github.com/bermannoah/metasploit-framework.git`. You might need to do it in a different directory than your version of the repo.

2 - cd into the repo directory

3 - just in case, run `git fetch` to make sure you have all branches. 

4 - `git checkout lhost-setting-warning` && `git pull origin lhost-setting-warning` (just in case)

5 - you'll need to go into the `core.rb` file and remove or alter my old solution to this issue

6 - make the changes you have in mind. make sure they work etc etc

7 - do all the usual `git add .` and `git commit -S -m "commit text here"` stuff. then run `git push origin lhost-setting-warning` and it should push to this PR. 

Let me know if that works for you or if you run into any troubles!

**Comment by itsmeroy2012** [comment]
When I run 'git checkout lhost-setting-warning && git pull origin lhost-setting-warning ' there is a fatal error. Will this be a problem later?

screenshot from 2017-03-07 22-43-19

**Comment by itsmeroy2012** [comment]
And then there is this. I think I don't have the rights to write into the repository.
screenshot from 2017-03-07 22-50-10

**Comment by bermannoah** [comment]
@itsmeroy2012 oh shoot, I'm sorry. It looks like your branch is up to date so you won't need to run git pull. For the future though it looks like you're leaving out `origin` -- it should be `git pull origin branchname`. When it comes to the second command: it looks like you don't have a GPG key setup for your commits. The contributor guide for metasploit recommends it -- here's a [guide to generating a key](https://help.github.com/articles/generating-a-new-gpg-key/#platform-linux) and here's a [guide to connecting it to github](https://help.github.com/articles/signing-commits-using-gpg/).

Let me know if you have any more questions! Good luck! :)

**Comment by itsmeroy2012** [comment]
Everything ran successfully until this error popped out while executing the command "git push origin lhost-setting-warning"

remote: Permission to bermannoah/metasploit-framework.git denied to itsmeroy2012.
fatal: unable to access 'https://github.com/bermannoah/metasploit-framework.git/': The requested URL returned error: 403

**Comment by bermannoah** [comment]
@itsmeroy2012 if you run `git remote -v` what do you see?

**Comment by itsmeroy2012** [comment]
root@kali:~/bermannoah/metasploit-framework# git remote -v
origin	https://github.com/bermannoah/metasploit-framework.git (fetch)
origin	https://github.com/bermannoah/metasploit-framework.git (push)

**Comment by bermannoah** [comment]
@itsmeroy2012 awesome -- added you as a contributor to my repo. try again?

**Comment by itsmeroy2012** [comment]
Do I need to repeat the steps again? Because the same error pops out , the one I mentioned above.

**Comment by bermannoah** [comment]
@itsmeroy2012 did you accept the contributor invite?

**Comment by itsmeroy2012** [comment]
Thanks a lot for your precious time. I think it is done.

**Comment by bermannoah** [comment]
@itsmeroy2012 hooray! Glad we could get that figured out.

**Comment by itsmeroy2012** [comment]
Could you check if my solution is correct. I am a newbie, I might have done something wrong too.

**Comment by bermannoah** [comment]
let's get @busterb's feedback on it if they've got a sec! :)

**Comment by itsmeroy2012** [comment]
Alright, and is it possible to merge pull requests?

**Comment by rwhitcroft** [comment]
The updated code is not the same as what @busterb suggested, and there appears to be some style issues and at least one syntax error (`==` vs `===`).

Additionally, HTTP(S) transports aren't handled - please also update `reverse_http.rb`.

**Comment by itsmeroy2012** [comment]
@rwhitcroft  ... Should I update both the reverse.rb and the reverse_http.rb file? And what are the style issues involved and also the syntax error  (== vs ===) ...  Where am I supposed to change it?

**Comment by rwhitcroft** [comment]
Yes, both files need to be updated. Your code should look identical to the code @busterb suggested here: https://github.com/rapid7/metasploit-framework/pull/8063/files#r104604972 - watch your spacing, indentation, etc.

It might also be helpful to get more familiar with the fundamentals of Git and Ruby before diving right into making PRs. Just a suggestion!

**Comment by bermannoah** [comment]
Hi @rwhitcroft @busterb -- I went ahead and made the changes as requested. I can confirm that the warning appears as needed and ... doesn't appear when it isn't. 

screen shot 2017-03-09 at 11 31 26 am

screen shot 2017-03-09 at 11 33 28 am

It also works for all of the test cases @busterb listed in his code suggestion. Let me know if there's anything else you need from me / anything else I can do.

File: lib/msf/core/handler/reverse_http.rb [hunk 2]
```diff
@@ -356,6 +356,23 @@ def on_request(cli, req)
...
 
+
+          # Checking whether LHOST is a loopback address
+
+          def is_loopback_address(address)
+            begin
+              a = IPAddr.new(address.to_s)
+              return true if IPAddr.new('127.0.0.1/8') === a
+              return true if IPAddr.new('::1') === a
+              rescue
```
Comment (rwhitcroft) [inline lib/msf/core/handler/reverse_http.rb]: This `rescue` goes with the `begin` - un-ident.

File: lib/msf/core/handler/reverse_http.rb [hunk 3]
```diff
@@ -356,6 +356,23 @@ def on_request(cli, req)
...
+            end
+              return false
```
Comment (rwhitcroft) [inline lib/msf/core/handler/reverse_http.rb]: Un-indent.

File: lib/msf/core/handler/reverse_http.rb (hunk 2 above)
Comment (rwhitcroft) [inline lib/msf/core/handler/reverse_http.rb]: This `rescue` goes with the `begin` - un-indent.

File: lib/msf/core/handler/reverse_http.rb (hunk 3 above)
Comment (rwhitcroft) [inline lib/msf/core/handler/reverse_http.rb]: Un-indent this `return`.

File: lib/msf/core/handler/reverse_http.rb [hunk 4]
```diff
@@ -356,6 +356,23 @@ def on_request(cli, req)
...
+          end
+
+          if is_loopback_address(uri.host)
+            print_warning ("You are attempting to listen on a loopback address by setting LHOST to #{uri.host}, did you mean to set ReverseListenerBindAddress instead?\n")
```
Comment (rwhitcroft) [inline lib/msf/core/handler/reverse_http.rb]: Don't need the `\n`.

File: lib/msf/core/handler/reverse.rb [hunk 5]
```diff
@@ -82,12 +83,28 @@ def setup_handler
...
+            begin
+               a = IPAddr.new(address.to_s)
+               return true if IPAddr.new('127.0.0.1/8') === a
+               return true if IPAddr.new('::1') === a
+            rescue
+            end
+            return false
+          end
+
+           if is_loopback_address(ip)
```
Comment (rwhitcroft) [inline lib/msf/core/handler/reverse.rb]: Fix the indentation of this `if` block.

Comment (bermannoah) [inline lib/msf/core/handler/reverse.rb]: Sorry, jumped the gun a bit. This is fixed in my most recent commit. I'm pushing up a change that removes the newlines as well.

File: lib/msf/core/handler/reverse_http.rb (hunk 4 above)
Comment (bermannoah) [inline lib/msf/core/handler/reverse_http.rb]: Fixed both here and in `reverse.rb`. Cheers!

File: lib/msf/core/handler/reverse.rb (hunk 5 above)
Comment (bermannoah) [inline lib/msf/core/handler/reverse.rb]: Changes are up. Let me know + thanks again!

Comment (rwhitcroft) [inline lib/msf/core/handler/reverse.rb]: It looks good in `reverse_http.rb` but it's still screwy in `reverse.rb`.

Comment (bermannoah) [inline lib/msf/core/handler/reverse.rb]: @rwhitcroft Whoops. Fixed!

File: lib/msf/core/handler/reverse_http.rb [hunk 6]
```diff
@@ -356,6 +356,23 @@ def on_request(cli, req)
 
           print_status("Staging #{uuid.arch} payload (#{blob.length} bytes) ...")
 
+
+          # Checking whether LHOST is a loopback address
+
+          def is_loopback_address(address)
```
Comment (wvu-r7) [inline lib/msf/core/handler/reverse_http.rb]: Isn't this already defined by the included `Msf::Handler::Reverse`?

File: lib/msf/core/handler/reverse.rb [hunk 7]
```diff
@@ -49,6 +49,7 @@ def bind_addresses
         addrs
       end
 
+
```
Comment (wvu-r7) [inline lib/msf/core/handler/reverse.rb]: Extraneous whitespace.

File: lib/msf/core/handler/reverse_http.rb (hunk 6 above)
Comment (wvu-r7) [inline lib/msf/core/handler/reverse_http.rb]: Nevermind, it's in `bind_port`. You should move it.

File: lib/msf/core/handler/reverse.rb [hunk 8]
```diff
@@ -82,12 +83,28 @@ def setup_handler
                 'MsfPayload' => self,
                 'MsfExploit' => assoc_exploit
               })
+
+          #Checking whether LHOST is a loopback address
+
```
Comment (wvu-r7) [inline lib/msf/core/handler/reverse.rb]: Extraneous whitespace.

File: lib/msf/core/handler/reverse.rb [hunk 9]
```diff
@@ -82,12 +83,28 @@ def setup_handler
...
+               a = IPAddr.new(address.to_s)
+               return true if IPAddr.new('127.0.0.1/8') === a
+               return true if IPAddr.new('::1') === a
+            rescue
+            end
+            return false
+          end
+
+          if is_loopback_address(ip)
+		        print_warning ("You are attempting to listen on a loopback address by setting LHOST to #{ip}, did you mean to set ReverseListenerBindAddress instead?")
```
Comment (wvu-r7) [inline lib/msf/core/handler/reverse.rb]: Your indentation is wrong.

File: lib/msf/core/handler/reverse.rb [hunk 10]
```diff
@@ -82,12 +83,28 @@ def setup_handler
...
+	        end
+
           rescue
             ex = $!
             print_error("Handler failed to bind to #{ip}:#{local_port}:- #{comm} -")
           else
             ex = false
-
```
Comment (wvu-r7) [inline lib/msf/core/handler/reverse.rb]: Why?

File: lib/msf/core/handler/reverse.rb [hunk 11]
```diff
@@ -82,12 +83,28 @@ def setup_handler
                 'MsfPayload' => self,
                 'MsfExploit' => assoc_exploit
               })
+
+          #Checking whether LHOST is a loopback address
+
+          def is_loopback_address(address)
```
Comment (wvu-r7) [inline lib/msf/core/handler/reverse.rb]: Move this method to the module level so you can use it in `Msf::Handler::ReverseHttp`.

File: lib/msf/core/handler/reverse_http.rb (hunk 6 above)
Comment (wvu-r7) [inline lib/msf/core/handler/reverse_http.rb]: Remove this, since it'll be available in the included `Msf::Handler::Reverse` once you move it there.

**Comment by bermannoah** [comment]
Just pushed up changes per @wvu-r7's feedback!

**Comment by wvu-r7** [comment]
I deleted my comments. I wasn't reading the diff correctly. If the changes you made are accurate, keep them.

File: lib/msf/core/handler/reverse.rb [hunk 12]
```diff
@@ -82,12 +93,16 @@ def setup_handler
                 'MsfPayload' => self,
                 'MsfExploit' => assoc_exploit
               })
+
+          if is_loopback_address(ip)
```
Comment (wvu-r7) [inline lib/msf/core/handler/reverse.rb]: Please fix the indentation here. I had trouble reading the diff because it wasn't clear what went where.

Comment (bermannoah) [inline lib/msf/core/handler/reverse.rb]: Just pushed up the fix. Sorry for the confusion! I also kept the `is_loopback_address` method at the module level, that's a lot DRY'er than what I had before. Cheers!

File: lib/msf/core/handler/reverse.rb [hunk 13]
```diff
@@ -82,12 +93,16 @@ def setup_handler
                 'MsfPayload' => self,
                 'MsfExploit' => assoc_exploit
               })
+
+            if is_loopback_address(ip)
```
Comment (wvu-r7) [inline lib/msf/core/handler/reverse.rb]: It's still messed up. How are you editing the file?

Comment (bermannoah) [inline lib/msf/core/handler/reverse.rb]: What specifically needs to happen here? I've modified the indentation so it's not inline with `begin` -- is the issue the whitespace on 96? I'm using Atom to edit the file.

Comment (wvu-r7) [inline lib/msf/core/handler/reverse.rb]: https://github.com/bermannoah/metasploit-framework/blob/ef6831437a21d77565d66bb3eaad5aa350bfd457/lib/msf/core/handler/reverse.rb#L98

Comment (wvu-r7) [inline lib/msf/core/handler/reverse.rb]: I mean the `if` block, not just line 97.

Comment (bermannoah) [inline lib/msf/core/handler/reverse.rb]: @wvu-r7 Yikes. Apologies. Should be fixed now.

**Comment by wvu-r7** [comment]
How are there 48 comments on this??

File: lib/msf/core/handler/reverse.rb [hunk 14]
```diff
@@ -55,6 +55,17 @@ def bind_port
...
 
+      # Checking if LHOST is a loopback address
+      def is_loopback_address(address)
+        begin
+           a = IPAddr.new(address.to_s)
+           return true if IPAddr.new('127.0.0.1/8') === a
+           return true if IPAddr.new('::1') === a
+        rescue
+        end
+        return false
```
Comment (wvu-r7) [inline lib/msf/core/handler/reverse.rb]: Why did you add this `return`?

File: lib/msf/core/handler/reverse.rb (hunk 13 above)
Comment (wvu-r7) [inline lib/msf/core/handler/reverse.rb]: This methods needs to be named `is_loopback_address?`. Fixing myself.

File: lib/msf/core/handler/reverse.rb [hunk 15]
```diff
@@ -55,6 +55,17 @@ def bind_port
         (port > 0) ? port : datastore['LPORT'].to_i
       end
 
+      # Checking if LHOST is a loopback address
+      def is_loopback_address(address)
```
Comment (wvu-r7) [inline lib/msf/core/handler/reverse.rb]: This methods needs to be named `is_loopback_address?`. Fixing myself.

Comment (wvu-r7) [inline lib/msf/core/handler/reverse.rb]: This method needs to be named `is_loopback_address?`. Fixing myself.

File: lib/msf/core/handler/reverse.rb [hunk 16]
```diff
@@ -82,12 +93,16 @@ def setup_handler
                 'MsfPayload' => self,
                 'MsfExploit' => assoc_exploit
               })
+
+            if is_loopback_address(ip)
+              print_warning ("You are attempting to listen on a loopback address by setting LHOST to #{ip}, did you mean to set ReverseListenerBindAddress instead?")
```
Comment (wvu-r7) [inline lib/msf/core/handler/reverse.rb]: Extraneous whitespace and comma splice. Fixing.

**Comment by wvu-r7** [comment]
```
msf exploit(handler) > set lhost 192.168.33.1
lhost => 192.168.33.1
msf exploit(handler) > set reverselistenerbindaddress 127.0.0.1
reverselistenerbindaddress => 127.0.0.1
msf exploit(handler) > run

[!] You are attempting to listen on a loopback address by setting LHOST to 127.0.0.1, did you mean to set ReverseListenerBindAddress instead?
[*] Started reverse TCP handler on 127.0.0.1:4444 
[*] Starting the payload handler...
```

This does not work.

**Comment by busterb** [comment]
Yeah, that's exactly the opposite of what this was intended to alert on.

**Comment by itsmeroy2012** [comment]
When we give the command: -

 'set reverselistenerbindaddress 127.0.0.1' , the 'ip' variable in the file lib/msf/core/handler/reverse.rb  changes in the following code:-

 if is_loopback_address(ip)
              print_warning ("You are attempting to listen on a loopback address by setting LHOST to #{ip}, did you mean to set ReverseListenerBindAddress instead?")
            end

Why is this happening and does the same variable stores both the lhost and the reverselistenerbindaddress?

**Comment by wvu-r7** [comment]
@itsmeroy2012: It's looping through each bind address. Look at the code just above that.

**Comment by itsmeroy2012** [comment]
You're talking about this part right? @wvu-r7 

bind_addresses.each do |ip|
 begin ...
 ....

**Comment by wvu-r7** [comment]
@itsmeroy2012: Correct.

**Comment by itsmeroy2012** [comment]
What's the complexity of this issue now? Is it still under 'newbie-friendly' ? @wvu-r7

**Comment by itsmeroy2012** [comment]
This seems to solve the problem specified above. But there's one flaw, the user has to specify a value for LPORT else the callback_host  stores the value '172.16.176.36' for some reason. Should I push the change? @wvu-r7 


bind_addresses.each do |ip|
          begin
            self.listener_sock = Rex::Socket::TcpServer.create(
              'LocalHost' => ip,
              'LocalPort' => local_port,
              'Comm'      => comm,
              'Context'   =>
              {
                'Msf'        => framework,
                'MsfPayload' => self,
                'MsfExploit' => assoc_exploit
              })

            callback_host = datastore['LHOST']
            if is_loopback_address(callback_host)
              print_warning ("You are attempting to listen on a loopback address by setting LHOST to #{ip}, did you mean to set ReverseListenerBindAddress instead?")
            end

**Comment by wvu-r7** [comment]
Please see https://github.com/rapid7/metasploit-framework/commit/af3cd18c9fa656dfb2f5dd74686c30772779f945 for edification.

**Comment by itsmeroy2012** [comment]
What was the fix?? @wvu-r7

**Comment by wvu-r7** [comment]
Read the commit!!

File: lib/msf/core/handler/reverse.rb (hunk 15 above)
Comment (wvu-r7) [inline lib/msf/core/handler/reverse.rb]: And for the record, this code is @busterb's: https://github.com/rapid7/metasploit-framework/pull/8063#discussion_r104604972.

**Comment by pbarry-r7** [comment]
# Release Notes

Display warning message to users when use of loopback addresses may be unintentional (and might lead to undesired/unexpected behavior for the user).

**Comment by Alex942** [comment]
hey, i am stuck with this problem.. someone help me in solving this problem.......plzzzz

**Comment by wvu-r7** [comment]
@Alex942: What problem? Don't comment on old tickets for support. Go to IRC.

**Comment by Alex942** [comment]
@wvu-r7 my metasploit showing error when i am setting my lhost to 127.0.0.1......i am unable to fix this

**Comment by wvu-r7** [comment]
Not an error. It works fine, just not suggested.

**Comment by Alex942** [comment]
@wvu-r7 can you please explain me step by step how to fix it..........

**Comment by rwhitcroft** [comment]
@Alex942 You're not listening. This is not an error, so there's nothing to fix. You are telling Metasploit you want to listen on the local address (127.0.0.1) which means no clients will be able to connect to you. Metasploit thinks that's weird, so it shows you this warning.

Are you sure you understand what 127.0.0.1 means?

**Comment by Alex942** [comment]
@rwhitcroft i just want my metasploit to start a meterpreter session when i choose 127.0.0.1 as local host. That all.............i know you all are pro of this area and you know many times more than me thats why i am asking...............

**Comment by wvu-r7** [comment]
@Alex942: You have provided close to zero details about your problem. How can we help? This isn't even the right place still. Please ask on IRC. I don't want to lock this thread.

If you want to catch a shell on localhost, your payload needs to connect back to that address. Basically, `LHOST` should match between handler and payload. This is all the help I can provide without further information. Thanks.

**Comment by wvu-r7** [comment]
Here ya go: https://metasploit.help.rapid7.com/docs/joining-the-metasploit-irc-channel.

**Pull request merged**